src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/TempPE/TempPEBuildManager.cs

```diff
@@ -0,0 +1,550 @@
+  <Application private_binpath = ""{outputPath}""/>
+  <Assembly
+    codebase = ""{outputFileName}""
+    name = ""{moniker}""
+    version = ""0.0.0.0""
+    snapshot_id = ""1""
+    replaceable = ""True""
+  />
+</root>";
+        }
+
+        internal bool CompilationNeeded(HashSet<string> files, string outputFileName)
+        {
+            if (!_fileSystem.FileExists(outputFileName))
+                return true;
+
+            try
+            {
+                DateTime outputDateTime = _fileSystem.LastFileWriteTimeUtc(outputFileName);
+
+                foreach (string file in files)
+                {
+                    DateTime fileDateTime = _fileSystem.LastFileWriteTimeUtc(file);
+                    if (fileDateTime > outputDateTime)
+                        return true;
+                }
+            }
+            // if we can't read the file time of the output file, then we presumably can't compile to it either, so returning false is appropriate.
+            // if we can't read the file time of an input file, then we presumably can't read from it to compile either, so returning false is appropriate
+            catch (IOException)
+            { }
+            catch (UnauthorizedAccessException)
+            { }
+
+            return false;
+        }
+
+        private static string GetOutputFileName(string outputPath, string moniker)
+        {
+            // All monikers are project relative paths by defintion (anything else is a link), meaning the only invalid
+            // filename characters possible are path separators so we just replace them
+            return Path.Combine(outputPath, moniker.Replace('\\', '.') + ".dll");
+        }
+
+        protected virtual async Task CompileTempPEAsync(HashSet<string> filesToCompile, string outputFileName, CancellationToken token)
+        {
+            _telemetryService.PostProperty(TelemetryEventName.TempPECompilation, TelemetryPropertyName.TempPECompilationOutputFileName, outputFileName);
+            bool result = await _activeWorkspaceProjectContextHost.OpenContextForWriteAsync(accessor =>
+            {
+                return _compiler.CompileAsync(accessor.Context, outputFileName, filesToCompile, token);
+            });
+
+            // if the compilation failed or was cancelled we should clean up any old TempPE outputs lest a designer gets the wrong types, plus its what legacy did
+            if (!result)
+            {
+                try
+                {
+                    _fileSystem.RemoveFile(outputFileName);
+                }
+                catch (IOException)
+                { }
+                catch (UnauthorizedAccessException)
+                { }
+            }
+        }
+
+        protected virtual HashSet<string> GetFilesToCompile(string moniker, ImmutableHashSet<string> sharedInputs)
+        {
+            // This is a HashSet because we allow files to be both inputs and shared inputs, and we don't want to compile the same file twice
```

Comment:
I think I would add comment here about why you are adding shared inputs and the input together, ie some background.